### PR TITLE
Change the light attenuation formulas.

### PR DESCRIPTION
**Pull request by reduz**
-Much better looking, physically based.
-Almost negligible extra cost.

Pictures (Old Left, New Right):

image
image
image

Videos:

Old:

https://user-images.githubusercontent.com/6265307/103666003-d4bcfd00-4f52-11eb-8627-e5df584fd28b.mp4

New:

https://user-images.githubusercontent.com/6265307/103666034-db4b7480-4f52-11eb-894f-fe7f5d903c1c.mp4

### CODE REVIEW

**Comment by fire** [comment]
is it possible if you also changed the formulas in the gltf document for importing and export lights at the same time.

**Comment by akien-mga** [comment]
There are similar looking equations in other places, are they meant to change too?

(Non exhaustive list, I just looked for `spot_attenuation` and for some `pow(1.0 - x)` equations.)
```
modules/lightmapper_rd/lm_compute.glsl
322:                            attenuation *= pow(1.0 - d, lights.data[i].spot_attenuation);
servers/rendering/renderer_rd/shaders/giprobe_write.glsl
155:                    attenuation *= pow(1.0 - d, lights.data[light].spot_attenuation);
servers/rendering/renderer_rd/shaders/sdfgi_direct_light.glsl
211:                                    attenuation *= pow(1.0 - d, lights.data[i].spot_attenuation);
servers/rendering/renderer_rd/shaders/scene_forward.glsl
1222:   spot_attenuation *= 1.0 - pow(spot_rim, spot_att_angle.x);
servers/rendering/renderer_rd/shaders/volumetric_fog.glsl
280:                    float attenuation = pow(max(1.0 - d, 0.0), attenuation_energy.x);
336:                    float attenuation = pow(max(1.0 - d, 0.0), attenuation_energy.x);
```

**Review by clayjohn — COMMENTED** [review]
Looks good to me. Subject to Akien's comment above. It looks like you accidentally removed a line.

**Comment by reduz** [comment]
> There are similar looking equations in other places, are they meant to change too?

No, spot attenuation is ok, this will never be physically very correct anyway, we need to invert it though. @JFonS was planning to work on this, we can do it a separate PR, and this can be a good guide of all the places that need to be touched (though probably in this case, it needs to be inverted on the CPP)

**Comment by JFonS** [comment]
Yup, I already found a way to properly invert the angle attenuation. I will also make sure this attenuation is consistent across LM/SDFGI/GIProbe etc. because currently there are some discrepancies.

File: modules/lightmapper_rd/lm_compute.glsl [hunk 1]
```diff
@@ -249,6 +249,15 @@ float quick_hash(vec2 pos) {
 	return fract(sin(dot(pos * 19.19, vec2(49.5791, 97.413))) * 49831.189237);
 }
 
+float get_omni_attenuation(float distance, float inv_range, float decay) {
+	float nd = distance * inv_range;
+	nd *= nd;
+	nd *= nd; // nd^4
```
Comment (Calinou) [inline modules/lightmapper_rd/lm_compute.glsl:255]: Isn't this `nd^3`?

Comment (akien-mga) [reply modules/lightmapper_rd/lm_compute.glsl:255]: I thought this initially too but no, it's equivalent to:
```
float nd2 = nd * nd;
float nd4 = nd2 * nd2;
```

**Comment by reduz** [comment]
@clayjohn ok, added the missing fog one. Should probably be good to merge once CI passes.

**Review by clayjohn — APPROVED** [review]

**Pull request merged**